File: runme/mcgen/local/datan/solsrc/s4rn.c
#include <stdio.h>
#include <math.h>
#include <string.h>
#include "datsrc.h"
#include "grsrc.h"

/* Table of constant values */

static integer c__1 = 1, c__2 = 2, c__15 = 15, c__46 = 46, c__50 = 50;
static double c_dp5 = .5, c_dm1 = -1., c_d1 = 1.;

main(void)
{
    /* Initialized data */
    static double xa[2] = { 0.,0. }, xb[2] = { 10.,10. }, del = .1;
    static char capt[15] = "t = ***** dt   ";
    /* Local variables */
    static integer i, j, k, l, m, ihist, iswit, mt, nt, nws, nx;
    static double delphi, phi, his, rcs[46], dsq, rsn[46], xpl[46], ypl[46],
                  rand[2], epar[2], delx, dinv, hist[100],
                  xlst[100] /* was [2][50] */, xtry[2], d, uipar, h1[2], h2[2],
                  uiperp, v[100] /* was [2][50] */, w[2],
                  x[100] /* was [2][50] */, eperp[2], x0, dd, dx[2],
                  ui[2], vi[2], vj[2], uj[2];
    static char tx[15], ty[15];
    /* System generated locals */
    integer i__1, i__2;
    double d__1, d__2;
/* identify program to user */
    printf("%s\n", "Program S4RN performs");
    printf("%s\n", "Monte Carlo simulation of molecular motion");
    printf("%s\n\n", "and produces graphical output.");
/* initialization */
    d = 1.;
    dsq = d * d;
    delphi = .13962622222222221;
    phi = 0.;
    for (i = 0; i <= 45; ++i) {
	rcs[i] = cos(phi) * .5;
	rsn[i] = sin(phi) * .5;
	phi += delphi;
    }
/* let user choose */
    printf("%s\n", "Enter number of time steps (several 100 or 1000):");
    printf("%s", "> ");
    scanf("%li", &nt);
    mt = nt / 10;
/* ask for number of workstation */
    grnbws_();
    printf("%s\n", "Please, enter number of workstation:");
    printf("%s", "> ");
    scanf("%li", &nws);
    printf("%s\n", "Choose:");
    printf("%s\n", "1 - display of trajectories");
    printf("%s\n", "2 - display of histograms");
    printf("%s", "> ");
    scanf("%li", &iswit);
    (void)getchar();
    if (iswit == 1) {
/* initialize graphics for trajectories */
	gropen_();
	gropws_(&nws);
	grwncc_(xa, xb, &xa[1], &xb[1]);
	grfram_();
    } else if (iswit == 2) {
/* let user choose type of histogram */
        printf("%s\n", "Choose:");
        printf("%s\n", "1 - histogram of v_1");
        printf("%s\n", "2 - histogram of v_2");
        printf("%s\n", "3 - histogram of |v|");
        printf("%s\n", "4 - histogram of |v|^2 ");
        printf("%s", "> ");
        scanf("%li", &ihist);
        (void)getchar();
/* initialize histograms */
	nx = 100;
        memset(tx, ' ', 15);
        memset(ty, ' ', 15);
	if (ihist == 1) {
	    x0 = -1.;
	    delx = .02;
	    strncpy(tx, "v_1", 3);
	    strncpy(ty, "N(v_1#)", 7);
	} else if (ihist == 2) {
	    x0 = -1.;
	    delx = .02;
	    strncpy(tx, "v_2", 3);
	    strncpy(ty, "N(v_2#)", 7);
	} else if (ihist == 3) {
	    x0 = 0.;
	    delx = .01;
	    strncpy(tx, "v", 1);
	    strncpy(ty, "N(v)", 4);
	} else if (ihist == 4) {
	    x0 = 0.;
	    delx = .01;
	    strncpy(tx, "v^2", 3);
	    strncpy(ty, "N(v^2#)", 7);
	}
    }
/* choose initial positions */
    for (i = 1; i <= 50; ++i) {
L10:
	rnecuy_(rand, &c__2);
	for (k = 1; k <= 2; ++k) {
	    xtry[k - 1] = xa[k - 1] + .5 + rand[k - 1] *
                          (xb[k - 1] - xa[k - 1] - d);
	}
	if (i > 1) {
/* repeat if there is overlap with existing particle */
	    i__1 = i - 1;
	    for (j = 1; j <= i__1; ++j) {
/* Computing 2nd power */
		d__1 = xtry[0] - x[(j << 1) - 2];
/* Computing 2nd power */
		d__2 = xtry[1] - x[(j << 1) - 1];
		if (d__1 * d__1 + d__2 * d__2 <= dsq) {
		    goto L10;
		}
	    }
	}
	x[(i << 1) - 2] = xtry[0];
	x[(i << 1) - 1] = xtry[1];
	xlst[(i << 1) - 2] = x[(i << 1) - 2];
	xlst[(i << 1) - 1] = x[(i << 1) - 1];
/* draw initial positions */
	for (l = 1; l <= 46; ++l) {
	    xpl[l - 1] = x[(i << 1) - 2] + rcs[l - 1];
	    ypl[l - 1] = x[(i << 1) - 1] + rsn[l - 1];
	}
	i__1 = i % 8 + 1;
	grstcl_(&i__1);
	grplin_(&c__46, xpl, ypl);
	if (i == 1) {
/* draw double circle for sphere with non zero initial velocity */
	    for (l = 1; l <= 46; ++l) {
		xpl[l - 1] = x[(i << 1) - 2] + rcs[l - 1] * .9;
		ypl[l - 1] = x[(i << 1) - 1] + rsn[l - 1] * .9;
	    }
	    grplin_(&c__46, xpl, ypl);
	}
    }
/* set initial velocities */
    for (i = 2; i <= 50; ++i) {
	v[(i << 1) - 2] = 0.;
	v[(i << 1) - 1] = 0.;
    }
    rnecuy_(rand, &c__1);
    phi = rand[0] * 6.28318;
    v[0] = cos(phi) * 1.;
    v[1] = sin(phi) * 1.;
/* loop over all times */
    i__1 = nt;
    for (m = 1; m <= i__1; ++m) {
/* loop over all particles */
	for (i = 1; i <= 50; ++i) {
/* advance particle */
	    for (k = 1; k <= 2; ++k) {
		x[k + (i << 1) - 3] += v[k + (i << 1) - 3] * .01;
	    }
/* if advance is large enough draw piece of trajectory */
	    if (iswit == 1) {
		if ((d__1 = xlst[(i << 1) - 2] - x[(i << 1) - 2], abs(d__1)) 
			> del || (d__2 = xlst[(i << 1) - 1] - x[(i << 1) - 1],
			 abs(d__2)) > del) {
		    xpl[0] = xlst[(i << 1) - 2];
		    ypl[0] = xlst[(i << 1) - 1];
		    xpl[1] = x[(i << 1) - 2];
		    ypl[1] = x[(i << 1) - 1];
		    xlst[(i << 1) - 2] = x[(i << 1) - 2];
		    xlst[(i << 1) - 1] = x[(i << 1) - 1];
		    i__2 = i % 8 + 1;
		    grstcl_(&i__2);
		    grplin_(&c__2, xpl, ypl);
		}
	    }
	}
/* test for collisions */
	for (i = 1; i <= 50; ++i) {
/* test for collisions with wall */
	    for (k = 1; k <= 2; ++k) {
		if (x[k + (i << 1) - 3] < xa[k - 1] + .5 || x[k + (i << 1) - 3]
                    > xb[k - 1] - .5) {
		    v[k + (i << 1) - 3] = -v[k + (i << 1) - 3];
		}
	    }
/* test for collisions with another particle */
	    if (i < 50) {
		for (j = i + 1; j <= 50; ++j) {
		    dx[0] = x[(i << 1) - 2] - x[(j << 1) - 2];
		    dx[1] = x[(i << 1) - 1] - x[(j << 1) - 1];
/* Computing 2nd power */
		    d__1 = dx[0];
/* Computing 2nd power */
		    d__2 = dx[1];
		    dd = d__1 * d__1 + d__2 * d__2;
		    if (dd < dsq) {
			dinv = 1. / sqrt(dd);
			eperp[0] = dx[0] * dinv;
			eperp[1] = dx[1] * dinv;
			epar[0] = eperp[1];
			epar[1] = -eperp[0];
			mtxgcl_(v, vi, &c__2, &c__50, &i);
			mtxgcl_(v, vj, &c__2, &c__50, &j);
			mtxadv_(vi, vj, w, &c__2);
			mtxmsv_(w, w, &c_dp5, &c__2);
			mtxsbv_(vi, w, ui, &c__2);
			mtxdot_(ui, eperp, &uiperp, &c__2);
			mtxdot_(ui, epar, &uipar, &c__2);
			uiperp = -uiperp;
			mtxmsv_(eperp, h1, &uiperp, &c__2);
			mtxmsv_(epar, h2, &uipar, &c__2);
			mtxadv_(h1, h2, ui, &c__2);
			mtxadv_(ui, w, vi, &c__2);
			mtxmsv_(ui, uj, &c_dm1, &c__2);
			mtxadv_(uj, w, vj, &c__2);
			mtxpcl_(v, vi, &c__2, &c__50, &i);
			mtxpcl_(v, vj, &c__2, &c__50, &j);
		    }
		}
	    }
	}
	if (iswit == 2) {
	    if (m % mt == 1 || m == nt) {
/* produce histogram */
                sprintf(capt + 4, "%5li", m);
                capt[9] = ' ';
		smhsin_(hist, &x0, &delx, &nx);
		for (j = 1; j <= 50; ++j) {
		    if (ihist == 1) {
			his = v[(j << 1) - 2];
		    }
		    if (ihist == 2) {
			his = v[(j << 1) - 1];
		    }
		    if (ihist == 3) {
/* Computing 2nd power */
			d__1 = v[(j << 1) - 2];
/* Computing 2nd power */
			d__2 = v[(j << 1) - 1];
			his = sqrt(d__1 * d__1 + d__2 * d__2);
		    }
		    if (ihist == 4) {
/* Computing 2nd power */
			d__1 = v[(j << 1) - 2];
			his = d__1 * d__1 + v[(j << 1) - 1];
		    }
		    smhsfl_(hist, &x0, &delx, &nx, &his, &c_d1);
		}
		smhsgr_(hist, &x0, &delx, &nx, tx, &c__15, ty, &c__15, capt,
			 &c__15, &nws);
	    }
	}
    }
    grclse_();
    return 0;
} /* main */

File: runme/mcgen/local/datan/solsrc/s4rn.f
      PROGRAM S4RN
      IMPLICIT DOUBLE PRECISION(A-H,O-Z)
      PARAMETER(N=50,R=.5D0,DT=1.D-2,V0=1.D0,PI=3.14159D0)
      DIMENSION XA(2),XB(2),X(2,N),XLST(2,N),DX(2),V(2,N),XTRY(2)
      DIMENSION VI(2),VJ(2),W(2),UI(2),UJ(2),H1(2),H2(2)
      DIMENSION EPERP(2),EPAR(2)
      DIMENSION RAND(2)
      DIMENSION XPL(46),YPL(46)
      DIMENSION HIST(100)
      DIMENSION RCS(46),RSN(46)
      CHARACTER*15 TX,TY,CAPT
      DATA XA/0.D0,0.D0/,XB/10.D0,10.D0/
      DATA DEL/.1D0/
      DATA CAPT/'t = ***** dt'/
C identify program to user
      WRITE(*,*)' Program S4RN performs'
      WRITE(*,*)' Monte Carlo simulation of molecular motion'
      WRITE(*,*)' and produces graphical output'
      WRITE(*,*)' '
C initialization
      D=2.D0*R
      DSQ=D*D
      DELPHI=2.D0*PI/45.D0
      PHI=0.D0
      DO 5 I=0,45
        RCS(I+1)=R*COS(PHI)
        RSN(I+1)=R*SIN(PHI)
        PHI=PHI+DELPHI
5     CONTINUE
C let user choose
      WRITE(*,*)'Enter number of time steps (several 100 or 1000)'
      WRITE(*,*)' >'
      READ(*,*)NT
      MT=NT/10
C ask for number of workstation
      CALL GRNBWS()
      WRITE(*,*)' Please, enter number of workstation:'
      WRITE(*,*)' >'
      READ(*,*)NWS
      WRITE(*,*)' Choose'
      WRITE(*,*)' 1 - display of trajectories'
      WRITE(*,*)' 2 - display of histograms'
      WRITE(*,*)' >'
      READ(*,*)ISWIT
      IF(ISWIT.EQ.1) THEN
C initialize graphics for trajectories
        CALL GROPEN
        CALL GROPWS(NWS)
        CALL GRWNCC(XA(1),XB(1),XA(2),XB(2))
        CALL GRFRAM
      ELSE IF(ISWIT.EQ.2) THEN
C let user choose type of histogram
        WRITE(*,*)' Choose'
        WRITE(*,*)' 1 - histogram of v_1'
        WRITE(*,*)' 2 - histogram of v_2'
        WRITE(*,*)' 3 - histogram of |v|'
        WRITE(*,*)' 4 - histogram of |v|^2 '
        WRITE(*,*)' >'
        READ(*,*)IHIST
C initialize histograms
        NX=100
        IF(IHIST.EQ.1)THEN
          X0=-V0
          DELX=0.02D0*V0
          TX='v_1'
          TY='N(v_1#)'
        ELSE IF(IHIST.EQ.2)THEN
          X0=-V0
          DELX=0.02D0*V0
          TX='v_2'
          TY='N(v_2#)'
        ELSE IF(IHIST.EQ.3) THEN
          X0=0.D0
          DELX=0.01D0*V0
          TX='v'
          TY='N(v)'
        ELSE IF(IHIST.EQ.4) THEN
          X0=0.D0
          DELX=0.01D0*V0**2
          TX='v^2'
          TY='N(v^2#)'
        END IF
      END IF
C choose initial positions
      DO 40 I=1,N
10      CALL RNECUY(RAND,2)
        DO 20 K=1,2
          XTRY(K)=XA(K)+R+RAND(K)*(XB(K)-XA(K)-D)
20      CONTINUE
        IF(I.GT.1) THEN
C repeat if there is overlap with existing particle
          DO 30 J=1,I-1
            IF((XTRY(1)-X(1,J))**2+(XTRY(2)-X(2,J))**2.LE.DSQ)
     +      GO TO 10
30        CONTINUE
        END IF
        X(1,I)=XTRY(1)
        X(2,I)=XTRY(2)
        XLST(1,I)=X(1,I)
        XLST(2,I)=X(2,I)
C draw initial positions
        DO 35 L=1,46
          XPL(L)=X(1,I)+RCS(L)
          YPL(L)=X(2,I)+RSN(L)
35      CONTINUE
        CALL GRSTCL(MOD(I,8)+1)
        CALL GRPLIN(46,XPL,YPL)
        IF(I.EQ.1) THEN
C draw double circle for sphere with non zero initial velocity
          DO 36 L=1,46
            XPL(L)=X(1,I)+.9D0*RCS(L)
            YPL(L)=X(2,I)+.9D0*RSN(L)
36        CONTINUE
          CALL GRPLIN(46,XPL,YPL)
        END IF
40    CONTINUE
C set initial velocities
      DO 45 I=2,N
        V(1,I)=0.D0
        V(2,I)=0.D0
45    CONTINUE
      CALL RNECUY(RAND,1)
      PHI=2.D0*PI*RAND(1)
      V(1,1)=V0*COS(PHI)
      V(2,1)=V0*SIN(PHI)
C loop over all times
      DO 1000 M=1,NT
C loop over all particles
        DO 60 I=1,N
C advance particle
          DO 50 K=1,2
            X(K,I)=X(K,I)+V(K,I)*DT
50        CONTINUE
C if advance is large enough draw piece of trajectory
          IF(ISWIT.EQ.1) THEN
            IF(ABS(XLST(1,I)-X(1,I)).GT.DEL .OR. ABS(XLST(2,I)-
     +      X(2,I)).GT.DEL) THEN
              XPL(1)=XLST(1,I)
              YPL(1)=XLST(2,I)
              XPL(2)=X(1,I)
              YPL(2)=X(2,I)
              XLST(1,I)=X(1,I)
              XLST(2,I)=X(2,I)
              CALL GRSTCL(MOD(I,8)+1)
              CALL GRPLIN(2,XPL,YPL)
            END IF
          END IF
60      CONTINUE
C test for collisions
        DO 200 I=1,N
C test for collisions with wall
          DO 70 K=1,2
            IF(X(K,I).LT.XA(K)+R.OR.X(K,I).GT.XB(K)-R) V(K,I)=-V(K,I)
70        CONTINUE
C test for collisions with another particle
          IF(I.LT.N) THEN
            DO 100 J=I+1,N
              DX(1)=X(1,I)-X(1,J)
              DX(2)=X(2,I)-X(2,J)
              DD=DX(1)**2+DX(2)**2
              IF(DD.LT.DSQ) THEN
                DINV=1.D0/SQRT(DD)
                EPERP(1)=DX(1)*DINV
                EPERP(2)=DX(2)*DINV
                EPAR(1)=EPERP(2)
                EPAR(2)=-EPERP(1)
                CALL MTXGCL(V,VI,2,N,I)
                CALL MTXGCL(V,VJ,2,N,J)
                CALL MTXADV(VI,VJ,W,2)
                CALL MTXMSV(W,W,0.5D0,2)
                CALL MTXSBV(VI,W,UI,2)
                CALL MTXDOT(UI,EPERP,UIPERP,2)
                CALL MTXDOT(UI,EPAR,UIPAR,2)
                UIPERP=-UIPERP
                CALL MTXMSV(EPERP,H1,UIPERP,2)
                CALL MTXMSV(EPAR,H2,UIPAR,2)
                CALL MTXADV(H1,H2,UI,2)
                CALL MTXADV(UI,W,VI,2)
                CALL MTXMSV(UI,UJ,-1.D0,2)
                CALL MTXADV(UJ,W,VJ,2)
                CALL MTXPCL(V,VI,2,N,I)
                CALL MTXPCL(V,VJ,2,N,J)
              END IF
100         CONTINUE
          END IF
200     CONTINUE
        IF(ISWIT.EQ.2) THEN
          IF(MOD(M,MT).EQ. 1.OR. M.EQ.NT) THEN
C produce histogram
            WRITE(CAPT(5:9),'(I5)')M
            CALL SMHSIN(HIST,X0,DELX,NX)
            DO 47 J=1,N
              IF(IHIST.EQ.1) HIS=V(1,J)
              IF(IHIST.EQ.2) HIS=V(2,J)
              IF(IHIST.EQ.3) HIS=SQRT(V(1,J)**2+V(2,J)**2)
              IF(IHIST.EQ.4) HIS=V(1,J)**2+V(2,J)
              CALL SMHSFL(HIST,X0,DELX,NX,HIS,1.D0)
47          CONTINUE
            CALL SMHSGR(HIST,X0,DELX,NX,TX,15,TY,15,CAPT,15,NWS)
          END IF
        END IF
1000  CONTINUE
      CALL GRCLSE
      END
